### PR TITLE
Support text records

### PR DESCRIPTION
**Pull request by makoto**
Please refer to https://github.com/ensdomains/ens-app/pull/1436

### CODE REVIEW

File: packages/client/src/index.ts [hunk 1]
```diff
@@ -82,6 +82,17 @@ const provider = new CCIPReadProvider(baseProvider);
...
+    const textData = Resolver.encodeFunctionData('text(bytes32,string)', [
+      node,
+      'description',
+    ]);
+    const textResponseData = await resolver.resolve(dnsName(name), textData);
+    const value = Resolver.decodeFunctionResult(
+      'text(bytes32,string)',
+      textResponseData
+    );
+    console.log(`${name}: ${value}`);
```
Comment (Arachnid) [inline packages/client/src/index.ts:95]: Should modify both of these log statements to state what they're logging - eg, 'addr()' and 'text(description)'.

File: packages/contracts/deploy/11_demo.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const { ethers } = require("hardhat");  
+module.exports = async ({getNamedAccounts, deployments, network}) => {
+    const {deploy} = deployments;
+    const {deployer, owner} = await getNamedAccounts();
+    var GATEWAY_HOST = 'https://offchain-resolver-example.uc.r.appspot.com'
```
Comment (Arachnid) [inline packages/contracts/deploy/11_demo.js:5]: Shouldn't this be using the export from arguments.js?

File: packages/contracts/arguments.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+module.exports = [
+  // Demo url
+  "https://offchain-resolver-example.uc.r.appspot.com/{sender}/{data}.json",
```
Comment (Arachnid) [inline packages/contracts/arguments.js:3]: I think these should probably be set as variables in hardhat.config.js - you can then access them from the network object in a deployment.

File: packages/contracts/deploy/11_demo.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const { ethers } = require("hardhat");
```
Comment (Arachnid) [inline packages/contracts/deploy/11_demo.js:1]: Shouldn't this modify the existing deployment to customise it for different networks? We don't need two deployments deploying the same thing.

File: packages/contracts/index.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+const IExtendedResolver = require('./artifacts/contracts/IExtendedResolver.sol/IExtendedResolver.json')
+const OffchainResolver = require('./artifacts/contracts/OffchainResolver.sol/OffchainResolver.json')
+
+module.exports = {
+  IExtendedResolver,
+  OffchainResolver
+}
```
Comment (Arachnid) [inline packages/contracts/index.js:7]: What is this needed for?

File: packages/contracts/package.json [hunk 6]
```diff
@@ -1,13 +1,14 @@
 {
   "name": "@ensdomains/offchain-resolver-contracts",
-  "version": "0.2.1",
+  "version": "0.0.2",
```
Comment (Arachnid) [inline packages/contracts/package.json:3]: ??

File: packages/gateway/package.json [hunk 7]
```diff
@@ -58,8 +58,8 @@
   "dependencies": {
     "@chainlink/ccip-read-server": "^0.2.1",
     "@ensdomains/ens-contracts": "^0.0.8",
-    "@ensdomains/offchain-resolver-contracts": "^0.2.1",
```
Comment (Arachnid) [inline packages/gateway/package.json:61]: Why is this being removed?

File: packages/gateway/src/index.ts [hunk 8]
```diff
@@ -3,20 +3,27 @@ import { Command } from 'commander';
...
-    'Private key to sign responses with. Prefix with @ to read from a file'
-  )
-  .requiredOption('-d --data <file>', 'JSON file to read data from')
-  .option('-t --ttl <number>', 'TTL for signatures', '300')
-  .option('-p --port <number>', 'Port number to serve on', '8000');
-
-program.parse(process.argv);
-
-const options = program.opts();
+import 'dotenv/config';
```
Comment (Arachnid) [inline packages/gateway/src/index.ts:6]: What was wrong with using command-line options?

File: packages/gateway/src/json.ts [hunk 9]
```diff
@@ -40,6 +41,14 @@ export class JSONDatabase implements Database {
     return { addr: nameData.addresses[coinType], ttl: this.ttl };
   }
 
+  text(name: string, key: string) {
+    const nameData = this.findName(name);
+    if (!nameData || !nameData.text || !nameData.text[key]) {
+      return { value: ZERO_ADDRESS, ttl: this.ttl };
```
Comment (Arachnid) [inline packages/gateway/src/json.ts:47]: The value here should be the empty string.

File: packages/gateway/src/offchainexample.eth.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
+{
+    "offchainexample.eth": {
+        "addresses": {
```
Comment (Arachnid) [inline packages/gateway/src/offchainexample.eth.json:3]: The address and text records should probably be different between the wildcard and regular record, so people can verify which one they're getting.

File: packages/offchainresolver.eth.json [hunk 11]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (Arachnid) [inline packages/offchainresolver.eth.json:1]: I don't think this file should be here.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: packages/gateway/src/index.ts (hunk 8 above)
Comment (makoto) [reply packages/gateway/src/index.ts:6]: I can't deploy to app engine with command line

File: packages/contracts/arguments.js (hunk 3 above)
Comment (makoto) [reply packages/contracts/arguments.js:3]: `arguments.js` is a stand alone file which gets specified when running the command line so I don't think argument.js can read the environment variable from hardhat.config.js (such feature is not specified at https://hardhat.org/plugins/nomiclabs-hardhat-etherscan.html either)

```
npx hardhat verify --constructor-args arguments.js DEPLOYED_CONTRACT_ADDRESS

```

Comment (Arachnid) [reply packages/contracts/arguments.js:3]: You can set custom variables in the 'network' object for each network, and then access them from the network object inside the deployment.

File: packages/contracts/deploy/11_demo.js (hunk 4 above)
Comment (Arachnid) [reply packages/contracts/deploy/11_demo.js:1]: It would make more sense, then to make 10_offchain_resolver just deploy the resolver, and add a new deployment that sets subnodes etc and only runs on the hardhat network. We definitely shouldn't have two separate deployments that both deploy the same contract.

File: packages/contracts/index.js (hunk 5 above)
Comment (Arachnid) [reply packages/contracts/index.js:7]: This should be a typescript file, then, and maybe should pull in the typechain artefacts instead.

File: packages/gateway/src/index.ts (hunk 8 above)
Comment (Arachnid) [reply packages/gateway/src/index.ts:6]: You can: https://cloud.google.com/appengine/docs/standard/nodejs/runtime#application_startup

File: packages/contracts/index.js (hunk 5 above)
Comment (makoto) [reply packages/contracts/index.js:7]: Do you have any pointer/reference to how to do that?

Comment (Arachnid) [reply packages/contracts/index.js:7]: Looks like I was mistaken, and we don't have typechain set up in this repo right now. Let's just convert this to TS.

Comment (makoto) [reply packages/contracts/index.js:7]: `client` and `gateway` is `typescript` project but this `contract` part is not ts so all written in js. Isn't it strange to only convert this file to TS, or are you requesting me to convert the entire package to ts?

Comment (Arachnid) [reply packages/contracts/index.js:7]: You're right, my mistake. We should open an issue to add typechain later, though.

File: packages/contracts/deploy/10_offchain_resolver.js [hunk 12]
```diff
@@ -1,18 +1,18 @@
...
-    const resolver = await deploy('OffchainResolver', {
-        from: owner,
-        args: ['http://localhost:8000/{sender}/{data}.json', ['0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266']],
+    const {deployer, owner} = await getNamedAccounts();
+    if(network.name === 'hardhat'){
+        signer = deployer
+    }else{
+        signer = owner
+    }
+    if(!network.config.gatewayurl) throw("gatewayurl is missing on hardhat.config.js")
```
Comment (Arachnid) [inline packages/contracts/deploy/10_offchain_resolver.js:11]: Please put this in braces.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: packages/gateway/src/index.ts (hunk 8 above)
Comment (Arachnid) [reply packages/gateway/src/index.ts]: I don't understand, why can you not just leave this as-is for anyone running it using yarn, and include the arguments in app.yaml for the App Engine app?

File: packages/contracts/hardhat.config.js [hunk 13]
```diff
@@ -1,16 +1,79 @@
...
+if(process.env.DEPLOYER_KEY && process.env.OWNER_KEY) {
+  real_accounts = [process.env.DEPLOYER_KEY, process.env.OWNER_KEY];
+}
+const gatewayurl = "https://offchain-resolver-example.uc.r.appspot.com/{sender}/{data}.json"
 /**
  * @type import('hardhat/config').HardhatUserConfig
  */
+
+// Usage: INFURA_ID=$INFURA_ID ETHERSCAN_API_KEY=$ETHERSCAN_API_KEY npx hardhat  --network $NETWORK etherscanverify --resolver $CONTRACT_ADDRESS --signer $SIGNER_ADDRESS"
+task("etherscanverify", "Verifies contracts")
```
[the remark this answers is not in the document]
Comment (makoto) [reply packages/contracts/hardhat.config.js:19]: yes, I tried but the default one doesn't let me pass gatewayurl from the network config.

**Comment by makoto** [comment]
https://github.com/ensdomains/offchain-resolver/pull/6#discussion_r809442107 what do you mean by "This still needs fixing"? You mean just to bump the number from 0.0.2 to 0.0.3?

File: packages/gateway/package.json [hunk 14]
```diff
@@ -11,7 +11,7 @@
     "node": ">=10"
   },
   "scripts": {
-    "start": "node dist/index.js",
+    "start": "node dist/index.js --private-key $PRIVATE_KEY --data $DATA --port $PORT",
```
[the remark this answers is not in the document]
Comment (Arachnid) [reply packages/gateway/package.json:14]: You should be able to interpolate environment variables directly into the command line inside app.yaml.

File: packages/contracts/hardhat.config.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (Arachnid) [reply packages/contracts/hardhat.config.js:19]: You don't have to; it verifies using the actual deployment parameters.

**Comment by makoto** [comment]
Are we using the same number (0.2.1) across all packages in the project (if that's the case, the root project is still 0.0.1)? I started from 0.0.1 because none of the packages have been published yet. Where did the number 0.2.1 come from? Even tag nor releases exist with the number.

```
~/.../ens/offchain-resolver (text)$egrep -e 'name|version' package.json packages/*/package.json
package.json:  "name": "@ensdomains/offchain-resolver",
package.json:  "version": "0.0.1",
packages/client/package.json:  "version": "0.2.1",
packages/client/package.json:  "name": "@ensdomains/offchain-resolver-client",
packages/contracts/package.json:  "name": "@ensdomains/offchain-resolver-contracts",
packages/contracts/package.json:  "version": "0.0.2",
packages/gateway/package.json:  "version": "0.2.1",
packages/gateway/package.json:  "name": "@ensdomains/offchain-resolver-gateway",
```

File: packages/contracts/hardhat.config.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (makoto) [reply packages/contracts/hardhat.config.js:19]: I get `Error HH305: Unrecognized param --apiurl` when I pass `npx hardhat --network mainnet etherscan-verify  --api-key $ETHERSCAN_API --apiurl  https://mainnet.infura.io/v3/INFURA_ID` and get `Error HH110: Invalid JSON-RPC response received: invalid project id` without it

[the remark this answers is not in the document]
Comment (Arachnid) [reply packages/contracts/hardhat.config.js:19]: You don't need `apiurl`; it will use the RPC URL from the hardhat config for the network you selected. You need to specify `INFURA_ID` on the command line so it's substituted into the config.

**Comment by Arachnid** [comment]
@makoto The root package is a meta-package for the workspace; it's not published anywhere, and so its version number doesn't really matter. I don't really mind what standard we use, but the version number can't go backwards.

File: README.md [hunk 15]
```diff
@@ -32,11 +32,10 @@ Next, build and run the gateway:
 
 ```
 cd packages/gateway
-yarn && yarn build
-yarn start --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 --data test.eth.json
+yarn && yarn build && yarn start --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 --data test.eth.json --port 8080
 ```
 
-The value for the `--private-key` flag should be the key you generated earlier.
+The value for the `PRIVATE_KEY` envrionment key should be the key you generated earlier.
```
Comment (Arachnid) [inline README.md:38]: This needs reverting.

File: packages/contracts/deploy/10_offchain_resolver.js [hunk 16]
```diff
@@ -1,18 +1,20 @@
...
-module.exports = async ({deployments}) => {
+module.exports = async ({getNamedAccounts, deployments, network}) => {
     const {deploy} = deployments;
-    const signers = await ethers.getSigners();
-    const owner = signers[0].address;
-    const resolver = await deploy('OffchainResolver', {
-        from: owner,
-        args: ['http://localhost:8000/{sender}/{data}.json', ['0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266']],
+    const {deployer, owner} = await getNamedAccounts();
+    if(!network.tags.demo){
```
Comment (Arachnid) [inline packages/contracts/deploy/10_offchain_resolver.js:6]: You can avoid the need for this by setting the appropriate account indexes for test networks in hardhat.config.js.

File: packages/gateway/package.json [hunk 17]
```diff
@@ -58,8 +58,9 @@
   "dependencies": {
     "@chainlink/ccip-read-server": "^0.2.1",
     "@ensdomains/ens-contracts": "^0.0.8",
-    "@ensdomains/offchain-resolver-contracts": "^0.2.1",
+    "@ensdomains/offchain-resolver-contracts": "^0.0.2",
```
Comment (Arachnid) [inline packages/gateway/package.json:61]: This is still set to 0.0.2.

File: packages/gateway/src/index.ts [hunk 18]
```diff
@@ -3,7 +3,7 @@ import { Command } from 'commander';
 import { readFileSync } from 'fs';
 import { ethers } from 'ethers';
 import { JSONDatabase } from './json';
-
+let options: any = {};
```
Comment (Arachnid) [inline packages/gateway/src/index.ts:6]: Why is this here?

File: packages/gateway/src/server.ts [hunk 19]
```diff
@@ -1,9 +1,11 @@
 import { Server } from '@chainlink/ccip-read-server';
-import { abi as IResolverService_abi } from '@ensdomains/offchain-resolver-contracts/artifacts/contracts/OffchainResolver.sol/IResolverService.json';
-import { abi as Resolver_abi } from '@ensdomains/ens-contracts/artifacts/contracts/resolvers/Resolver.sol/Resolver.json';
 import { ethers, BytesLike } from 'ethers';
 import { hexConcat, Result } from 'ethers/lib/utils';
-
+import { ETH_COIN_TYPE } from './utils';
+const IResolverService_abi = require('@ensdomains/offchain-resolver-contracts/artifacts/contracts/OffchainResolver.sol/IResolverService.json')
```
Comment (Arachnid) [inline packages/gateway/src/server.ts:5]: These should be imports.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: packages/contracts/deploy/10_offchain_resolver.js (hunk 16 above)
Comment (makoto) [reply packages/contracts/deploy/10_offchain_resolver.js]: I just flipped the order of deployer and owner so that owner is always `accounts[0]` which shouldn't break any other dependencies.

File: packages/gateway/src/server.ts (hunk 19 above)
Comment (makoto) [reply packages/gateway/src/server.ts:5]: I don't know why but I get the following when I try to import

```
Could not find a declaration file for module '@ensdomains/offchain-resolver-contracts'. '/Users/makoto/work/ens/offchain-resolver/packages/gateway/node_modules/@ensdomains/offchain-resolver-contracts/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/ensdomains__offchain-resolver-contracts` if it exists or add a new declaration (.d.ts) file containing `declare module '@ensdomains/offchain-resolver-contracts';`ts(7016)
```

File: README.md [hunk 20]
```diff
@@ -32,8 +32,7 @@ Next, build and run the gateway:
 
 ```
 cd packages/gateway
-yarn && yarn build
-yarn start --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 --data test.eth.json
+yarn && yarn build && yarn start --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 --data test.eth.json --port 8080
```
Comment (Arachnid) [inline README.md:35]: Is this change necessary any longer?

Comment (Arachnid) [inline README.md:35]: Is this change necessary any longer?

**Review by Arachnid — APPROVED** [review]
